### DISCUSSION TITLE
Set an environment variable without restarting the container

### DISCUSSION THREAD

**Discussion by MurzNN** [ddev/ddev#4658] [Q&A] [answered]
Is there any way to set or change an env (environment variable) value without the full restart of the web container?

I know that I can temporarily set it for the current terminal, but need to set it for other processes (including the php-fpm) too.

And it's ok that it will be dropped or reset after the restart.

**Comment by MurzNN** [comment]
I've invented a workaround to set the ENV variable to the `php-fpm` process and current term via a simple one-liner:
```
printf "\nenv[BOTVA] = botven\n" >> /etc/php/$DDEV_PHP_VERSION/fpm/pool.d/www.conf; supervisorctl restart php-fpm
```
Maybe it's a good idea to move this to a utility script inside `ddev`? Do we have them already?

**Comment by MurzNN** [comment, marked as answer]
And here is a more detailed script, that applies the ENV value to the new terminals, and also to the current (if ran with `source` prefix):
`cat /usr/local/bin/ddev-setenv`:
```bash
#!/bin/bash

if [[ "$1" == *"="* ]]; then
  ENV=$(echo "$1" | cut -d '=' -f 1)
  VALUE=$(echo "$1" | cut -d '=' -f 2)
else
  ENV=$1
  VALUE=$2
fi

FILE_TERMINAL1=~/.profile
FILE_TERMINAL2=~/.bashrc
FILE_FPM=/etc/php/$DDEV_PHP_VERSION/fpm/pool.d/www.conf

# If arguments are empty.
if [ -z "$ENV" ]; then
  echo "Usage format:
$0 ENV value
or
$0 ENV=value
"
  exit
fi

# Removing the line if the value is empty, because php-fpm doesn't support empty values.
if [ -z "$VALUE" ]; then
  sed -i "s/^env\[$ENV\] = .*//" $FILE_FPM
else
  sed -i "/^env\[$ENV\] = /{h;s/ = .*/ = $VALUE/};\${x;/^$/{s//env\[$ENV\] = $VALUE/;H};x}" $FILE_FPM
fi

sed -i "/^export $ENV=/{h;s/=.*/=$VALUE/};\${x;/^$/{s//export $ENV=$VALUE/;H};x}" $FILE_TERMINAL1
sed -i "/^export $ENV=/{h;s/=.*/=$VALUE/};\${x;/^$/{s//export $ENV=$VALUE/;H};x}" $FILE_TERMINAL2

supervisorctl restart php-fpm

# This will not work if ran as a subshell
# The solution is to run the script with `source` prefix
# source /usr/local/bin/setenv NAME VALUE
export $ENV="$VALUE"
if [ "$0" = "$BASH_SOURCE" ]; then
  echo "Warning: To apply the ENV variable also to the current terminal, execute it via this command:"
  echo "source $0 $1 $2"
fi

echo "Environment variable successfully set: $ENV=\"$VALUE\""
```

Usage example:
```
$ source ddev-setenv MY_COOL_VARIABLE "My cool value"
php-fpm: stopped
php-fpm: started
Environment variable successfully set: MY_COOL_VARIABLE=My cool value
```

Can we include it into DDEV web container?

**Reply by rfay** [reply]
There are a few questions that have to be answered and understood..

1. What .env are you talking about? There is `.ddev/.env`, which is read and set by docker-compose during startup. And then there are several usages by various CMSs. Some of them use environment variables set in php-fpm, but most read application-level .env files (usually in docroot) when they run, so are not dependent on on what php-fpm sees. 
2. You seem to be intent on changing environment variables known by php-fpm, which will not be either of the above. There is a new way to set those in v1.21.5-alpha1 and upcoming v1.21.5, but I'm not sure it affects what you're trying to do. But `.ddev/web-entrypoint.d/*.sh` scripts can do things including setting environment variables there. 

What class of the above are you wanting to work on?

**Reply by rfay** [reply]
Sorry, I initially read `.env`, so my answer is off-target.

To get your script into the container, just add it to your project and use it there. For example, put it in the project's `bin` directory, which is automatically added to $PATH.

**Reply by MurzNN** [reply]
> To get your script into the container, just add it to your project and use it there. For example, put it in the project's `bin` directory, which is automatically added to $PATH.

Yes, I know how to mount it manually, but I'm asking about including it (or something similar) into the DDEV package, because I think it's a pretty common task to change an ENV variable value quickly, without a full slow restart of the container. What do you think about it?

_For me, this script is very useful to test feature flags in the app, so via a couple of seconds I can switch it on and off, and test the behavior of the app._

_And not all apps read use `.env` file to read values, most of them just take it from the environment directly, so writing to `.env` file is not a solution._

**Reply by rfay** [reply]
You're the first to ask this in all these years, probably because most apps don't depend on env variables being defined in php-fpm, they use their own techniques to get the environment. What CMS or framework are you working with ? Maybe if I understand more clearly the problem you're trying to solve I can be more helpful.

**Reply by MurzNN** [reply]
I'm working with Drupal, but "feature flags" is a common practice for many projects. So, when you're developing and releasing a new feature to production, you put a dependency on an env variable, eg `ENABLE_ROUTING_V2=1` to enable it, so if the brand new feature works not such good as expected, you can quickly disable it, and revert the previous behavior via `ENABLE_ROUTING_V2=0`. 

So to test this locally - developers need a way to quickly change the ENV variable value, without waiting for full restart of the container.

And my script is already delivered to our team, and all are happy with it :)

**Reply by rfay** [reply]
So if you were using the [dotenv](https://www.drupal.org/project/dotenv) module for vars like this you could just change the .env and it would take immediate effect, right? Also note that you can accomplish the same thing by just changing a settings.local.php file, which would be the most common Drupal way to handle things like this.

**Reply by MurzNN** [reply]
Yes, using the `dotenv` is a workaround, but it is a redundant dependency for projects that work in containers, because it does an additional check of the file on each query. But `settings.local.php` is a good workaround, thanks for the tip! But this script should work for all scripts, not only for Drupal.
